File: variant-a/java-service/src/main/java/com/flashsale/api/service/AdaptiveInventoryV2.java
package com.flashsale.api.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.RedisTemplate;

import java.math.BigDecimal;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Adaptive Inventory V2 - Producer-Consumer Pattern with Async Refills
 *
 * Corrected implementation for Variant A flash sale:
 * - Each allocation unit starts with allocated_quantity items in RAM (e.g., 500)
 * - When hitting low water mark (e.g., 30% = 150 items), trigger async refill
 * - Async refill pulls refill_batch_size items (e.g., 50) from Redis campaign pool
 * - Refills cascade: 500→150→45→13... (each refill hits new 30% threshold)
 * - Three-tier fallback: Local RAM → Redis campaign pool → Ordinary stock
 */
public class AdaptiveInventoryV2 {

    private static final Logger logger = LoggerFactory.getLogger(AdaptiveInventoryV2.class);

    // Allocation unit configuration
    private final Long allocationId;
    private final UUID campaignId;
    private final UUID skuId;
    private final int allocatedQuantity;     // Initial quantity (e.g., 500)
    private final int refillBatchSize;        // Refill amount (e.g., 50)
    private final BigDecimal lowWaterMarkPct; // e.g., 0.30 (30%)

    // Redis client
    private final RedisTemplate<String, Object> redisTemplate;
    private final String campaignPoolKey;  // fs:{campaignId}:limit

    // Local stock management (Producer-Consumer)
    private final AtomicInteger localStock;
    private final AtomicInteger currentLowWaterMark;
    private final ReentrantLock stockLock = new ReentrantLock();

    // Refill management (single-flight pattern)
    private final AtomicBoolean refillInProgress = new AtomicBoolean(false);

    // Metrics
    private final AtomicLong totalRequests = new AtomicLong(0);
    private final AtomicLong ramHits = new AtomicLong(0);
    private final AtomicLong redisRefills = new AtomicLong(0);
    private final AtomicLong campaignPoolHits = new AtomicLong(0);

    public AdaptiveInventoryV2(
            Long allocationId,
            UUID campaignId,
            UUID skuId,
            int allocatedQuantity,
            int refillBatchSize,
            BigDecimal lowWaterMarkPct,
            RedisTemplate<String, Object> redisTemplate) {

        this.allocationId = allocationId;
        this.campaignId = campaignId;
        this.skuId = skuId;
        this.allocatedQuantity = allocatedQuantity;
        this.refillBatchSize = refillBatchSize;
        this.lowWaterMarkPct = lowWaterMarkPct;
        this.redisTemplate = redisTemplate;

        // Initialize local stock with allocated quantity
        this.localStock = new AtomicInteger(allocatedQuantity);

        // Calculate initial low water mark (30% of allocated quantity)
        int initialLowWaterMark = (int) (allocatedQuantity * lowWaterMarkPct.doubleValue());
        this.currentLowWaterMark = new AtomicInteger(initialLowWaterMark);

        // Redis key for campaign pool
        this.campaignPoolKey = String.format("fs:%s:limit", campaignId);

        logger.info(
            "[Allocation {}] Initialized: {} items in RAM, low_water_mark={}, refill_batch={}",
            allocationId, allocatedQuantity, initialLowWaterMark, refillBatchSize
        );
    }

    /**
     * Reserve one item using producer-consumer pattern
     *
     * Returns: ReservationResult with success flag, price type, and price value
     */
    public ReservationResult reserveItem() {
        totalRequests.incrementAndGet();

        // CONSUMER: Try local RAM first (fast path - pure memory)
        int current = localStock.get();
        if (current > 0) {
            if (localStock.compareAndSet(current, current - 1)) {
                ramHits.incrementAndGet();

                // Check if hit low water mark (trigger async refill)
                int newStock = current - 1;
                if (newStock == currentLowWaterMark.get() && !refillInProgress.get()) {
                    // Trigger async refill (PRODUCER - non-blocking!)
                    CompletableFuture.runAsync(this::doRefill);
                }

                return new ReservationResult(true, "campaign", BigDecimal.valueOf(79.99));
            }
            // CAS failed, retry
            return reserveItem();
        }

        // Local RAM depleted, try panic buffer (spin wait for refill)
        return handleDepleted();
    }

    /**
     * Handle depleted local stock
     * - Spin wait briefly (5-10ms) in case refill is in progress
     * - If still no stock, fall back to Redis campaign pool
     */
    private ReservationResult handleDepleted() {
        long spinStart = System.nanoTime();
        long spinDeadline = spinStart + 10_000_000; // 10ms max spin

        while (System.nanoTime() < spinDeadline) {
            Thread.onSpinWait(); // CPU hint for spin-wait

            int current = localStock.get();
            if (current > 0) {
                if (localStock.compareAndSet(current, current - 1)) {
                    ramHits.incrementAndGet();
                    return new ReservationResult(true, "campaign", BigDecimal.valueOf(79.99));
                }
            }

            // Check if should give up (no refill in progress and still 0)
            if (!refillInProgress.get() && localStock.get() == 0) {
                break;
            }
        }

        // Fallback: Try Redis campaign pool directly
        return tryRedisCampaignPool();
    }

    /**
     * PRODUCER: Execute async refill operation
     * Pulls refill_batch_size items from Redis campaign pool to local RAM
     */
    private void doRefill() {
        // Single-flight protection
        if (!refillInProgress.compareAndSet(false, true)) {
            return; // Another thread is already refilling
        }

        long refillStart = System.nanoTime();

        try {
            // Calculate refill amount
            int refillAmount = refillBatchSize;

            // Check if refill amount too small (not worth overhead)
            int currentStock = localStock.get();
            if (currentStock < 10 && refillAmount < 10) {
                logger.debug(
                    "[Allocation {}] Skipping refill: stock={}, refill={} too small",
                    allocationId, currentStock, refillAmount
                );
                return;
            }

            // DECRBY Redis campaign pool (ONLY HERE - rare Redis call!)
            Long campaignRemaining = redisTemplate.opsForValue().decrement(campaignPoolKey, refillAmount);

            if (campaignRemaining != null && campaignRemaining >= 0) {
                // Success: Add to local stock
                stockLock.lock();
                try {
                    int oldStock = localStock.get();
                    int newStock = oldStock + refillAmount;
                    localStock.set(newStock);

                    // Update cascading low water mark (30% of new stock)
                    int newLowWaterMark = (int) (newStock * lowWaterMarkPct.doubleValue());
                    currentLowWaterMark.set(newLowWaterMark);

                    redisRefills.incrementAndGet();

                    long refillLatencyMs = (System.nanoTime() - refillStart) / 1_000_000;

                    logger.info(
                        "[Allocation {}] Refill SUCCESS: {} → {} (+{}), " +
                        "new_low_mark={}, campaign_pool={}, latency={}ms",
                        allocationId, oldStock, newStock, refillAmount,
                        newLowWaterMark, campaignRemaining, refillLatencyMs
                    );
                } finally {
                    stockLock.unlock();
                }
            } else {
                // Campaign pool depleted, rollback
                redisTemplate.opsForValue().increment(campaignPoolKey, refillAmount);
                logger.warn(
                    "[Allocation {}] Refill FAILED: campaign pool depleted",
                    allocationId
                );
            }

        } catch (Exception e) {
            logger.error("[Allocation " + allocationId + "] Refill error", e);
        } finally {
            refillInProgress.set(false);
        }
    }

    /**
     * Try to reserve from Redis campaign pool directly (fallback tier 2)
     */
    private ReservationResult tryRedisCampaignPool() {
        Long remaining = redisTemplate.opsForValue().decrement(campaignPoolKey);

        if (remaining != null && remaining >= 0) {
            campaignPoolHits.incrementAndGet();
            logger.debug(
                "[Allocation {}] Reserved from campaign pool, remaining={}",
                allocationId, remaining
            );
            return new ReservationResult(true, "campaign", BigDecimal.valueOf(79.99));
        } else {
            // Restore if went negative
            if (remaining != null && remaining < 0) {
                redisTemplate.opsForValue().increment(campaignPoolKey);
            }

            // Campaign pool exhausted, fall back to ordinary stock (tier 3)
            logger.warn(
                "[BENCHMARK STOP INDICATOR] Allocation {} campaign pool exhausted, " +
                "falling back to ordinary stock",
                allocationId
            );
            return tryOrdinaryStock();
        }
    }

    /**
     * Try to reserve from ordinary stock (fallback tier 3)
     * In benchmark mode, this should stop the benchmark (campaign exhausted)
     */
    private ReservationResult tryOrdinaryStock() {
        String ordinaryStockKey = String.format("inv:%s", skuId);
        Long remaining = redisTemplate.opsForValue().decrement(ordinaryStockKey);

        if (remaining != null && remaining >= 0) {
            logger.warn(
                "[Allocation {}] Using ordinary stock (price=199.99), campaign exhausted",
                allocationId
            );
            return new ReservationResult(true, "ordinary", BigDecimal.valueOf(199.99));
        } else {
            // Restore if went negative
            if (remaining != null && remaining < 0) {
                redisTemplate.opsForValue().increment(ordinaryStockKey);
            }

            // Completely sold out
            return new ReservationResult(false, "sold_out", BigDecimal.ZERO);
        }
    }

    /**
     * Get current metrics for this allocation unit
     */
    public InventoryMetrics getMetrics() {
        return new InventoryMetrics(
            allocationId,
            localStock.get(),
            currentLowWaterMark.get(),
            totalRequests.get(),
            ramHits.get(),
            redisRefills.get(),
            campaignPoolHits.get(),
            refillInProgress.get()
        );
    }

    /**
     * Reservation result containing success flag, price type, and price
     */
    public static class ReservationResult {
        public final boolean success;
        public final String priceType;  // "campaign", "ordinary", "sold_out"
        public final BigDecimal price;

        public ReservationResult(boolean success, String priceType, BigDecimal price) {
            this.success = success;
            this.priceType = priceType;
            this.price = price;
        }
    }

    /**
     * Metrics for an allocation unit
     */
    public static class InventoryMetrics {
        public final Long allocationId;
        public final int localStock;
        public final int currentLowWaterMark;
        public final long totalRequests;
        public final long ramHits;
        public final long redisRefills;
        public final long campaignPoolHits;
        public final boolean refillInProgress;

        public InventoryMetrics(
                Long allocationId, int localStock, int currentLowWaterMark,
                long totalRequests, long ramHits, long redisRefills,
                long campaignPoolHits, boolean refillInProgress) {
            this.allocationId = allocationId;
            this.localStock = localStock;
            this.currentLowWaterMark = currentLowWaterMark;
            this.totalRequests = totalRequests;
            this.ramHits = ramHits;
            this.redisRefills = redisRefills;
            this.campaignPoolHits = campaignPoolHits;
            this.refillInProgress = refillInProgress;
        }

        public double getRamHitRate() {
            return totalRequests > 0 ? (double) ramHits / totalRequests * 100 : 0;
        }
    }
}

File: variant-a/python-service/app/services/adaptive_inventory_v2.py
"""
Corrected Adaptive Inventory Service for Variant A
Producer-Consumer Pattern with Cascading Low Water Marks

Design:
- Local memory counter (pure RAM, no Redis on hot path)
- Async refill on low water mark (30% threshold)
- Cascading thresholds (150 → 45 → 13...)
- SpinWait panic buffer (brief wait if refill in progress)
- Semaphore coordination (single-flight refill)
"""

import asyncio
import time
from typing import Optional
import redis.asyncio as redis
import logging

logger = logging.getLogger(__name__)


class AdaptiveInventoryUnit:
    """
    Single allocation unit with producer-consumer pattern

    Consumer: Request threads decrement local_stock
    Producer: Async refill thread replenishes from Redis campaign pool
    """

    def __init__(
        self,
        allocation_id: int,
        campaign_id: str,
        sku_id: str,
        allocated_quantity: int,
        refill_batch_size: int,
        low_water_mark_pct: float,
        redis_client: redis.Redis
    ):
        self.allocation_id = allocation_id
        self.campaign_id = campaign_id
        self.sku_id = sku_id
        self.redis = redis_client

        # Configuration (from DB allocation table)
        self.initial_quantity = allocated_quantity
        self.refill_batch_size = refill_batch_size
        self.low_water_mark_pct = low_water_mark_pct / 100.0  # 30.00 → 0.30

        # Local memory counter (CONSUMER reads/writes)
        self._local_stock = allocated_quantity
        self._stock_lock = asyncio.Lock()

        # Refill coordination (PRODUCER)
        self._refill_semaphore = asyncio.Semaphore(1)  # Single-flight refill
        self._refill_in_progress = False
        self._current_low_water_mark = int(allocated_quantity * self.low_water_mark_pct)

        # Metrics
        self._total_requests = 0
        self._ram_hits = 0  # Served from local memory
        self._redis_refills = 0  # Times refilled from Redis
        self._redis_direct_hits = 0  # Fell back to Redis campaign pool

        # Redis keys
        # Refill from SKU-specific pool (Partitioned Layer 2)
        self._campaign_pool_key = f"fs:{campaign_id}:redis_pool:sku:{sku_id}"
        self._ordinary_stock_key = f"inv:{sku_id}"  # Ordinary stock fallback

        logger.info(
            f"[Allocation {allocation_id}] Initialized: {allocated_quantity} items, "
            f"refill_batch={refill_batch_size}, low_water_mark={self._current_low_water_mark}, "
            f"refill_source={self._campaign_pool_key}"
        )

    async def reserve_item(self) -> tuple[bool, str, float]:
        """
        Reserve one item using producer-consumer pattern

        Returns:
            (success, price_type, price_value)
            price_type: 'campaign', 'ordinary', or 'sold_out'
        """
        self._total_requests += 1

        # Fast path: Try local memory (CONSUMER - pure RAM)
        async with self._stock_lock:
            if self._local_stock > 0:
                self._local_stock -= 1
                self._ram_hits += 1

                # Check if hit low water mark (trigger async refill)
                if self._local_stock == self._current_low_water_mark and not self._refill_in_progress:
                    # Trigger async refill (non-blocking)
                    asyncio.create_task(self._async_refill())

                return (True, 'campaign', 79.99)

        # Slow path: Local depleted, try panic buffer
        return await self._handle_depleted()

    async def _handle_depleted(self) -> tuple[bool, str, float]:
        """
        Handle local stock depletion

        Flow:
        1. If refill in progress, SpinWait briefly (panic buffer)
        2. Try Redis campaign pool (dual stock)
        3. Fall back to ordinary stock
        """

        # Panic Buffer: If refill in progress, wait briefly for it to complete
        if self._refill_in_progress:
            logger.debug(f"[Allocation {self.allocation_id}] Panic buffer: refill in progress, waiting...")

            # SpinWait up to 10ms for refill to complete
            start = time.perf_counter()
            max_wait_ms = 10

            while (time.perf_counter() - start) * 1000 < max_wait_ms:
                async with self._stock_lock:
                    if self._local_stock > 0:
                        # Refill completed! Take from newly refilled stock
                        self._local_stock -= 1
                        self._ram_hits += 1
                        logger.debug(f"[Allocation {self.allocation_id}] Panic buffer SUCCESS: got item from refill")
                        return (True, 'campaign', 79.99)

                await asyncio.sleep(0.001)  # 1ms spin

            logger.debug(f"[Allocation {self.allocation_id}] Panic buffer timeout, falling back to Redis")

        # Try Redis campaign pool (DUAL STOCK - still campaign price)
        campaign_remaining = await self.redis.decr(self._campaign_pool_key)

        if campaign_remaining >= 0:
            self._redis_direct_hits += 1
            logger.info(
                f"[Allocation {self.allocation_id}] Redis campaign pool hit, "
                f"remaining: {campaign_remaining}"
            )
            return (True, 'campaign', 79.99)
        else:
            # Campaign sold out, rollback the decrement
            await self.redis.incr(self._campaign_pool_key)

        # Fall back to ordinary stock (BENCHMARK SHOULD STOP HERE)
        ordinary_remaining = await self.redis.decr(self._ordinary_stock_key)

        if ordinary_remaining >= 0:
            logger.warning(
                f"[Allocation {self.allocation_id}] Fell back to ordinary stock! "
                f"Campaign sold out. Ordinary remaining: {ordinary_remaining}"
            )
            return (True, 'ordinary', 99.99)
        else:
            # Completely sold out
            await self.redis.incr(self._ordinary_stock_key)
            return (False, 'sold_out', 0.0)

    async def _async_refill(self):
        """
        PRODUCER: Async refill from Redis campaign pool

        Uses semaphore to prevent concurrent refills (single-flight pattern)
        Implements cascading low water marks
        """

        # Try to acquire semaphore (non-blocking check)
        if not self._refill_semaphore.locked():
            async with self._refill_semaphore:
                await self._do_refill()
        else:
            logger.debug(f"[Allocation {self.allocation_id}] Refill already in progress, skipping")

    async def _do_refill(self):
        """Execute the actual refill operation"""

        self._refill_in_progress = True
        refill_start = time.perf_counter()

        try:
            # Calculate refill amount
            refill_amount = self.refill_batch_size

            # Check if refill amount too small (not worth overhead)
            current_stock = self._local_stock
            if current_stock < 10 and refill_amount < 10:
                logger.info(
                    f"[Allocation {self.allocation_id}] Skipping refill: "
                    f"stock={current_stock}, refill={refill_amount} too small"
                )
                return

            # DECRBY Redis campaign pool (ONLY HERE - rare Redis call!)
            campaign_remaining = await self.redis.decrby(self._campaign_pool_key, refill_amount)

            if campaign_remaining >= 0:
                # Success: Add to local stock
                async with self._stock_lock:
                    old_stock = self._local_stock
                    self._local_stock += refill_amount

                    # Update cascading low water mark (30% of new stock)
                    self._current_low_water_mark = int(self._local_stock * self.low_water_mark_pct)

                    self._redis_refills += 1

                    refill_latency_ms = (time.perf_counter() - refill_start) * 1000

                    logger.info(
                        f"[Allocation {self.allocation_id}] Refill SUCCESS: "
                        f"{old_stock} → {self._local_stock} (+{refill_amount}), "
                        f"new_low_mark={self._current_low_water_mark}, "
                        f"campaign_pool={campaign_remaining}, "
                        f"latency={refill_latency_ms:.2f}ms"
                    )
            else:
                # Campaign pool depleted, rollback
                await self.redis.incrby(self._campaign_pool_key, refill_amount)
                logger.warning(
                    f"[Allocation {self.allocation_id}] Refill FAILED: "
                    f"campaign pool depleted"
                )

        except Exception as e:
            logger.error(f"[Allocation {self.allocation_id}] Refill error: {e}")

        finally:
            self._refill_in_progress = False

    def get_metrics(self) -> dict:
        """Get performance metrics"""
        return {
            'allocation_id': self.allocation_id,
            'local_stock': self._local_stock,
            'current_low_water_mark': self._current_low_water_mark,
            'total_requests': self._total_requests,
            'ram_hits': self._ram_hits,
            'redis_refills': self._redis_refills,
            'redis_direct_hits': self._redis_direct_hits,
            'ram_hit_rate': f"{(self._ram_hits / self._total_requests * 100) if self._total_requests > 0 else 0:.2f}%"
        }

    async def get_stranded_stock(self) -> int:
        """Get remaining local stock (for write-back on campaign end)"""
        async with self._stock_lock:
            return self._local_stock

    async def clear_stock(self):
        """Clear local stock (after write-back)"""
        async with self._stock_lock:
            self._local_stock = 0


class AdaptiveInventoryManager:
    """
    Manages multiple allocation units for a campaign/SKU
    Each service instance may claim multiple units
    """

    def __init__(self, redis_client: redis.Redis):
        self.redis = redis_client
        self._units: dict[int, AdaptiveInventoryUnit] = {}  # allocation_id → unit
        self._sku_units: dict[str, list[AdaptiveInventoryUnit]] = {}  # sku_id → [units]
        self._round_robin_index: dict[str, int] = {}  # sku_id → current index

    def add_unit(
        self,
        allocation_id: int,
        campaign_id: str,
        sku_id: str,
        allocated_quantity: int,
        refill_batch_size: int,
        low_water_mark_pct: float
    ):
        """Add an allocation unit to be managed"""

        unit = AdaptiveInventoryUnit(
            allocation_id=allocation_id,
            campaign_id=campaign_id,
            sku_id=sku_id,
            allocated_quantity=allocated_quantity,
            refill_batch_size=refill_batch_size,
            low_water_mark_pct=low_water_mark_pct,
            redis_client=self.redis
        )

        self._units[allocation_id] = unit

        if sku_id not in self._sku_units:
            self._sku_units[sku_id] = []
            self._round_robin_index[sku_id] = 0

        self._sku_units[sku_id].append(unit)

        logger.info(f"Added allocation unit {allocation_id} for SKU {sku_id}")

    async def reserve_item(self, sku_id: str) -> tuple[bool, str, float]:
        """
        Reserve item from any available unit for this SKU
        Uses round-robin across units for load balancing
        """

        if sku_id not in self._sku_units or not self._sku_units[sku_id]:
            # No local units - fall back to Redis directly
            logger.info(f"No allocation units for SKU {sku_id}, falling back to Redis campaign pool")
            return await self._fallback_to_redis(sku_id)

        units = self._sku_units[sku_id]
        start_index = self._round_robin_index[sku_id]

        # Try each unit in round-robin order
        for i in range(len(units)):
            index = (start_index + i) % len(units)
            unit = units[index]

            success, price_type, price = await unit.reserve_item()

            if success and price_type == 'campaign':
                # Update round-robin for next request
                self._round_robin_index[sku_id] = (index + 1) % len(units)
                return (success, price_type, price)

        # All units depleted, try first unit's fallback logic (Redis campaign pool + ordinary)
        return await units[0]._handle_depleted()

    async def _fallback_to_redis(self, sku_id: str) -> tuple[bool, str, float]:
        """
        Fallback to Redis when no local units available
        Tries: campaign pool → ordinary stock
        """
        # Assume campaign_id and keys (would need to be passed or configured)
        campaign_pool_key = f"fs:750e8400-e29b-41d4-a716-446655440000:limit"
        ordinary_stock_key = f"inv:{sku_id}"

        # Try Redis campaign pool
        campaign_remaining = await self.redis.decr(campaign_pool_key)
        if campaign_remaining >= 0:
            logger.info(f"Redis campaign pool hit for SKU {sku_id}, remaining: {campaign_remaining}")
            return (True, 'campaign', 79.99)
        else:
            # Rollback
            await self.redis.incr(campaign_pool_key)

        # Fall back to ordinary stock
        ordinary_remaining = await self.redis.decr(ordinary_stock_key)
        if ordinary_remaining >= 0:
            logger.warning(f"Fell back to ordinary stock for SKU {sku_id}, remaining: {ordinary_remaining}")
            return (True, 'ordinary', 99.99)
        else:
            # Completely sold out
            await self.redis.incr(ordinary_stock_key)
            return (False, 'sold_out', 0.0)

    def get_all_metrics(self) -> list[dict]:
        """Get metrics from all units"""
        return [unit.get_metrics() for unit in self._units.values()]

    async def get_total_stranded_stock(self, sku_id: str) -> int:
        """Get total stranded stock across all units for a SKU"""
        if sku_id not in self._sku_units:
            return 0

        total = 0
        for unit in self._sku_units[sku_id]:
            total += await unit.get_stranded_stock()

        return total

    async def clear_all_stock(self, sku_id: str):
        """Clear all stock for a SKU (after write-back)"""
        if sku_id in self._sku_units:
            for unit in self._sku_units[sku_id]:
                await unit.clear_stock()
